Add option to group cameras from sessions

justinvshen opened the pull request:
### Description
This new feature allows users to group different camera angles into distinct groups, where upon exporting a training set, the user cab export all frames from a specific group.

### Types of changes

- [ ] Bugfix
- [X] New feature
- [ ] Refactor / Code style update (no logical changes)
- [ ] Build / CI changes
- [ ] Documentation Update
- [ ] Other (explain)

### Does this address any currently open issues?
[list open issues here]

### Outside contributors checklist

- [ ] Review the [guidelines for contributing](https://github.com/talmolab/sleap/blob/develop/docs/CONTRIBUTING.md) to this repository
- [ ] Read and sign the [CLA](https://github.com/talmolab/sleap/blob/develop/sleap-cla.pdf) and add yourself to the [authors list](https://github.com/talmolab/sleap/blob/develop/AUTHORS)
- [ ] Make sure you are making a pull request against the **develop** branch (not *main*). Also you should start *your branch* off *develop*
- [ ] Add tests that prove your fix is effective or that your feature works
- [ ] Add necessary documentation (if appropriate)

#### Thank you for contributing to SLEAP!
:heart:

Jvshen committed:
Imports

Jvshen committed:
Add a button to add camera to group

Jvshen committed:
Add camera to group function

Jvshen committed:
Create Camera Groups table model

Jvshen committed:
Create camera groups table

Jvshen committed:
Create camera groups table

Jvshen committed:
Add camera groups table model to class attributes

Jvshen committed:
Add camera groups table function and adding and deleteing a camera group function

Jvshen committed:
Add table model for camera groups

roomrys commented on sleap/gui/dataviews.py:
```diff
@@ -683,6 +683,40 @@ def item_to_data(self, obj: RecordingSession, item: Camcorder):
+        if key == "name" and value:
+            item.name = value
+            # Mark project as changed
+            self.context.changestack_push("rename camera group")
```
While we want to call a `self.context` method here, this is not the one (we haven't created the method we want to call yet).
```suggestion
            # TODO(JS): Add `CommandContext` method that updates the camera group name
            raise NotImplementedError
```

roomrys commented on sleap/gui/dataviews.py:
```diff
@@ -683,6 +683,40 @@ def item_to_data(self, obj: RecordingSession, item: Camcorder):
+        """Update the model when camera groups change."""
+        self.items = camera_groups
+        self.beginResetModel()
+        self.endResetModel()
```
Oh man, yeah I do kinda like this better than what we currently do:
https://github.com/talmolab/sleap/blob/653144760882aa3f5c7f1199366d45aba0bccaed/sleap/gui/app.py#L1216-L1220
. Let's leave it for now (and possibly open a new PR to refactor all the other table updates).

roomrys commented on sleap/gui/widgets/docks.py:
```diff
@@ -625,10 +630,52 @@ def create_video_unlink_button(self) -> QWidget:
+                "No Camera Selected",
+                "Please select a camera to add to a group."
+            )
+            return
```
Instead of raising an error message, maybe we can just disable the button so no one can click on it if there is no camera selected. See this example for the "Link Video" button:
https://github.com/talmolab/sleap/blob/d898a8487669d70479117c6ddd22cdc9bd2035ba/sleap/gui/app.py#L1191-L1195
.
```suggestion
```

Jvshen committed:
Merge branch 'liezl/add-gui-elements-for-sessions' of https://github.com/talmolab/sleap into justin/group-cameras-from-sessions

Jvshen committed:
Update sessions dock to be able to add and delete camera groups and add a camera to a camera group

Jvshen committed:
Initialize camera_groups state variables in App class

Jvshen committed:
Update on_data_update to sync camera_groups between labels metadata and app state

Jvshen committed:
Add setCameraGroupName method to enable renaming camera groups in table

Jvshen committed:
Add CameraGroup class, AddCameraGroup, DeleteCameraGroup,
SetCameraGroupName, AddCameraToGroup, and RemoveCameraFromGroup for creating, deleting, and managing camera groups

Jvshen committed:
Add camera group methods to CommandContext for executing camera group commands

roomrys commented on sleap/gui/dataviews.py:
```diff
@@ -683,6 +683,40 @@ def item_to_data(self, obj: RecordingSession, item: Camcorder):
+    properties = ("name", "cameras")
+    
+    def __init__(self, items=None, context=None):
+        super().__init__(items=items, context=context)
```
Expose the property kwarg:
```suggestion
    def __init__(self, items=None, properties=None, context=None):
        super().__init__(items=items, properties=properties, context=context)
```

roomrys commented on sleap/gui/widgets/docks.py:
```diff
@@ -736,6 +800,124 @@ def lay_everything_out(self) -> None:
+            self.main_window, "New Camera Group", "Enter name for camera group:"
+        )
+        if ok and name:
+            self.main_window.commands.addCameraGroup(name)
```
```suggestion
            # TODO(JS): self.main_window.commands.addCameraGroup(name)
            raise NotImplementedError
```

roomrys commented on sleap/gui/widgets/docks.py:
```diff
@@ -625,10 +630,52 @@ def create_video_unlink_button(self) -> QWidget:
+            return
+        
+        # Get available groups
+        if not self.main_window.state.get("camera_groups") or len(self.main_window.state["camera_groups"]) == 0:
```
`GuiState` has a `__getitem__` method, so we at least won't run into an error if the key doesn't exist.
https://github.com/talmolab/sleap/blob/653144760882aa3f5c7f1199366d45aba0bccaed/sleap/gui/state.py#L52-L54

How about we use `get`, but with a default of an empty list:
```suggestion
        if len(self.main_window.state.get("camera_groups", []) == 0:
```

roomrys replied:
ex:
```python
In [20]: from sleap.gui.state import GuiState

In [21]: state = GuiState()

In [22]: state[3]

In [23]: state.get(3, [])
Out[23]: []
```

roomrys commented on sleap/gui/widgets/docks.py:
```diff
@@ -736,6 +800,124 @@ def lay_everything_out(self) -> None:
+        if ok and selected_name:
+            selected_idx = group_names.index(selected_name)
+            selected_group = self.main_window.state["camera_groups"][selected_idx]
+            self.main_window.commands.addCameraToGroup(camera, selected_group)
```
This entire method is a duplicate
```suggestion
```

roomrys reviewed (changes requested):
A few things:

1. If a method does not exist, instead of calling the method that doesn't exist, raise a `NotImplementedError` and add a `TODO(JS)`.
2. Also, can we see some screenshots of the GUI elements in the PR. 
3. I would also like to see tests that hits each one of the methods you've created.

roomrys commented on sleap/gui/commands.py:
```diff
@@ -3918,3 +3936,131 @@ def do_action(context: CommandContext, params: dict):
+    
+    def remove_camera(self, camera):
+        if camera in self.cameras:
+            self.cameras.remove(camera)
```
We should rename this class to something like `CameraCategory` and then place it in the same file as the existing `CameraGroup`, `InstanceGroup`, `FrameGroup` classes.

Jvshen committed:
Change naming and remove unnecessary code

Jvshen committed:
Add import

Jvshen committed:
Refactor: Rename function CameraGroup to CameraCategory in CommandContext

Jvshen committed:
Refactor: Rename function CameraGroup to CameraCategory in command classes

Jvshen committed:
Refactor: Rename function CameraGroupsTableModel to CameraCategoriesTableModel

Jvshen committed:
Refactor: Rename function CameraGroup to CameraCategory in docks

Jvshen committed:
Add CameraCategory class to cameras.py

Jvshen committed:
Add camera_categories attribute

Jvshen committed:
Add YAML config for Export Labels Package dialog

Jvshen committed:
Add menu item for Export Labels Package dialog

Jvshen committed:
Add ExportLabelsPackage command for handling dialog-based exports

Jvshen committed:
Add execute function for ExportLabelsPackage in CommandContext

Jvshen committed:
Add dialog function for collecting export options

Jvshen committed:
Move export labels dialog into command class and remove export_labels.py

Jvshen committed:
Format change

Jvshen committed:
Add Camera_category option to exporting

Jvshen committed:
formatting

Jvshen committed:
Make ExportLabelsPackage be a subclass of ExportDatasetWithImages

Jvshen committed:
Spacing and formatting

Jvshen committed:
Dialog for exporting labels

Jvshen committed:
make_cattr for cameraCategory

Jvshen committed:
Filter labels to include frames from cameras in camera categories

Jvshen committed:
Testing to see if labels store and retrieve camera categories correctly

roomrys commented on sleap/gui/app.py:
```diff
@@ -995,6 +997,16 @@ def new_instance_menu_action():
+            "Export labeled frames with various options into a single file.\n\n"
+            "Use this for archiving, training, or sharing your dataset."
+        )
+
```
Let's also remove the existing menu item(s) as this replaces all the export labels functionality.

roomrys commented on sleap/gui/commands.py:
```diff
@@ -1555,7 +1586,66 @@ class ExportFullPackage(ExportDatasetWithImages):
+        # Show the export labels dialog and return the selected options
+        dialog = FormBuilderModalDialog(form_name="export_labels_form", parent=parent)
+        dialog.setWindowTitle("Export Labels Package")
+        return dialog.get_results()
```
```suggestion
```

roomrys commented on sleap/gui/commands.py:
```diff
@@ -63,6 +63,8 @@ class which inherits from `AppCommand` (or a more specialized class such as
 from sleap.skeleton import Node, Skeleton
 from sleap.util import get_package_file
+from sleap.io.cameras import CameraCategory
+from sleap.gui.dialogs.formbuilder import FormBuilderModalDialog
```
```suggestion
```

roomrys commented on sleap/gui/commands.py:
```diff
@@ -1555,7 +1586,66 @@ class ExportFullPackage(ExportDatasetWithImages):
+            suggested=params["suggested"],
+            camera_category=params.get("camera_category"),
+            verbose=True,
+        )
```
If subclassing ExportDatasetWithImages, we can remove this.
```suggestion
```

roomrys commented on sleap/gui/commands.py:
```diff
@@ -1555,7 +1586,66 @@ class ExportFullPackage(ExportDatasetWithImages):
     suggested = True
 
 
-# Navigation Commands
```
Add back this comment below your new class.

roomrys commented on sleap/gui/commands.py:
```diff
@@ -3918,3 +4008,152 @@ def do_action(context: CommandContext, params: dict):
+
+        # Initialize camera_categories if needed
+        if not hasattr(context.labels, "camera_categories"):
+            context.labels.camera_categories = []
```
We can remove because Labels.camera_categories now always exists.
```suggestion
```

roomrys commented on sleap/gui/commands.py:
```diff
@@ -3918,3 +4008,152 @@ def do_action(context: CommandContext, params: dict):
+        context.labels.camera_categories.append(camera_category)
+
+        # Update state
+        context.state["camera_categories"] = context.labels.camera_categories
```
I'm not completely sure why we update the state here? Is it accessed/used somewhere? Why not just access it through `context.state["labels"].camera_categories`?

roomrys commented on sleap/gui/commands.py:
```diff
@@ -3918,3 +4008,152 @@ def do_action(context: CommandContext, params: dict):
+                context.labels.camera_categories.remove(camera_category)
+
+                # Update state
+                context.state["camera_categories"] = context.labels.camera_categories
```
Not sure we need to set the `"camera_categories"` state.

Jvshen committed:
add camera category filter to export labels dialog

Jvshen committed:
remove old export labels package button

Jvshen committed:
remove unused import

Jvshen committed:
remove extra do_action method

Jvshen committed:
Add back navigation commands comment

Jvshen committed:
remove since Labels.camera_categories now always exists

Jvshen committed:
Set, add, and populate combobox with available camera categories

Jvshen committed:
Formatting

Jvshen committed:
Remove unused method

Jvshen committed:
Remove "camera_categories"

Jvshen committed:
Change to category

Jvshen committed:
Formatting

Jvshen committed:
WIP: Change add_camera_to_category workflow

Jvshen committed:
Formatting

roomrys committed:
Make changes to add camera to categories

roomrys committed:
Make changes to add camera to categories

roomrys committed:
Reset selection after command

roomrys committed:
Fix selected_camera_category not being selected

roomrys committed:
Link Delete Category button straight to command

roomrys committed:
Remove unused update model method

roomrys committed:
Move create category code to commands

roomrys committed:
Rename methods to more appropriate names

roomrys committed:
Disable delete category button when there is no selected category

roomrys committed:
Delegate code to commands instead of dock

roomrys committed:
Reorganize method in SessionsDock

Jvshen committed:
Add remove from category button

Jvshen committed:
Add defaults for removeCameraFromCateogory

Jvshen committed:
Raise error if camera category or camera is not selected

Jvshen committed:
Add remove from categoyr to update_gui_state

roomrys committed:
Delegate table update to command topic when rename

roomrys committed:
Rename variable

roomrys committed:
Add failing camera category load (or save) test

roomrys committed:
Redo to_dict and from_dict methods (standalone)

roomrys committed:
Test standalone to_dict and from_dict methods

roomrys committed:
Redo to and from dict for Labels integration

roomrys committed:
Typehint and docstring CameraCategory.make_cattr

roomrys committed:
Fix CameraCategory (de)serialization with Labels

roomrys committed:
Test CameraCategory (de)serialization with Labels

roomrys committed:
Lint

roomrys commented on sleap/gui/app.py:
```diff
@@ -1875,3 +1868,8 @@ def main(args: Optional[list] = None, labels: Optional[Labels] = None):
         app.exec_()
 
     pass
+
+
+if __name__ == "__main__":
+    ds = "tests/data/cameras/minimal_session/min_session_frame_groups.slp"
+    main([ds])
```
Remove before merging

roomrys committed:
[wip] Filter exported labels by camera category

roomrys committed:
[wip] Add test for export labels with category

Jvshen committed:
Add tests for executing exportlabelspackage and verifying cameras and categories were correctly created (WIP)
